examples/vision/metric_learning.py

```diff
@@ -0,0 +1,314 @@
+
+
+"""
+
+height_width = 32
+
+
+def show_collage(examples):
+    box_size = height_width + 2
+    num_rows, num_cols = examples.shape[:2]
+
+    collage = Image.new(
+        mode="RGB",
+        size=(num_cols * box_size, num_rows * box_size),
+        color=(250, 250, 250),
+    )
+    for row_idx in range(num_rows):
+        for col_idx in range(num_cols):
+            array = (np.array(examples[row_idx, col_idx]) * 255).astype(np.uint8)
+            collage.paste(
+                Image.fromarray(array), (col_idx * box_size, row_idx * box_size)
+            )
+
+    # Double size for visualisation.
+    collage = collage.resize((2 * num_cols * box_size, 2 * num_rows * box_size))
+    return collage
+
+
+# Show a collage of 5x5 random images.
+sample_idxs = np.random.randint(0, 50000, size=(5, 5))
+examples = x_train[sample_idxs]
+show_collage(examples)
+
+"""
+Metric learning provides training data not as explicit `(X, y)` pairs but instead uses
+multiple instances that are related in the way we want to express similarity. In our
+example we will use instances of the same class to represent similarity; a single
+training instance will not be one image, but a pair of images of the same class. When
+referring to the images in this pair we'll use the common metric learning names of the
+`anchor` (a randomly chosen image) and the `positive` (another randomly chosen image of
+the same class).
+
+"""
+
+"""
+To facilitate this we need to build a form of lookup that maps from classes to the
+instances of that class. When generating data for training we will sample from this
+lookup.
+"""
+
+class_idx_to_train_idxs = defaultdict(list)
+for y_train_idx, y in enumerate(y_train):
+    class_idx_to_train_idxs[y].append(y_train_idx)
+
+class_idx_to_test_idxs = defaultdict(list)
+for y_test_idx, y in enumerate(y_test):
+    class_idx_to_test_idxs[y].append(y_test_idx)
+
+"""
+For this example we are using the simplest approach to training; a batch will consist of
+`(anchor, positive)` pairs spread across the classes. The goal of learning will be to
+move the anchor and positive pairs closer together and further away from other instances
+in the batch. In this case the batch size will be dictated by the number of classes; for
+CIFAR-10 this is 10.
+"""
+
+num_classes = 10
+
+
+def anchor_positive_pairs_dataset(num_batchs):
+    def anchor_positive_pair():
+        for _ in range(num_batchs):
+            for class_idx in range(num_classes):
+                examples_for_class = class_idx_to_train_idxs[class_idx]
+                anchor_idx = random.choice(examples_for_class)
+                positive_idx = random.choice(examples_for_class)
+                while positive_idx == anchor_idx:
+                    positive_idx = random.choice(examples_for_class)
+                yield x_train[anchor_idx], x_train[positive_idx]
+
+    return (
+        tf.data.Dataset.from_generator(
```

Comment:
If you're going to use a generator, I recommend using the `Sequence` class, like here: https://keras.io/examples/vision/oxford_pets_image_segmentation/#prepare-sequence-class-to-load-amp-vectorize-batches-of-data

Reply:
done. but there's one quirk still, around slicing. see my other comment.